### PR TITLE
Fix for Oracle driver

### PR DESCRIPTION
**Pull request by klingac**
Adminer was unable to connect to Oracle DB.
Fixed editing CLOBs ina Oracle
Implemented insertUpdate method for Oracle DB

### CODE REVIEW

**Comment by helix84** [comment]
With this PR, trying to connect with SID in the Server field failed for me.

**Comment by klingac** [comment]
@helix84 try to use [Easy Connect String](https://docs.oracle.com/database/121/NETAG/naming.htm), e.g. 'localhost/xe'

**Comment by klingac** [comment]
I created docker image for dev and test with oracle OCI driver. Just checkout this PR and you can try that with command

    docker run -d --name=adminer -p 8080:80 -p 443:443 -v /path/to/adminer/repo:/var/www/html klingac/apache-php-oci

If you also need Oracle DB try to use any of theese: https://hub.docker.com/search/?isAutomated=0&isOfficial=0&page=1&pullCount=0&q=wnameless%2Foracle-xe-11g&starCount=0

Or use docker-compose.yml from this repo https://github.com/klingac/Apache-PHP-OCI
You only need to change 

`volumes:
    - ./html:/var/www/html
`

Adminer will be in [http:localhost:80/](http://localhost:80/adminer/adminer/?oracle=&username=system&password=oracle&db=oracledb%2Fxe)
Then use this credentials in Adminer (or use link above):

    host: null
    SID: oracledb/xe
    login: system
    password: oracle

File: adminer/drivers/oracle.inc.php [hunk 1]
```diff
@@ -121,12 +121,12 @@ class Min_DB extends Min_PDO {
...
 
 			function connect($server, $username, $password) {
-				$this->dsn("oci:dbname=//$server;charset=AL32UTF8", $username, $password);
+				$this->dsn("oci:dbname=$server;charset=AL32UTF8", $username, $password);
 				return true;
 			}
 
 			function select_db($database) {
-				return true;
+                return $database;
```
Comment (vrana) [inline adminer/drivers/oracle.inc.php]: What does this mean?

File: adminer/drivers/oracle.inc.php [hunk 2]
```diff
@@ -136,6 +136,41 @@ function select_db($database) {
 
 	class Min_Driver extends Min_SQL {
 
+        function insert($table, $set) {
+            return parent::insert($table, $set);
```
Comment (vrana) [inline adminer/drivers/oracle.inc.php]: What's the purpose of this?

File: adminer/drivers/oracle.inc.php [hunk 3]
```diff
@@ -136,6 +136,41 @@ function select_db($database) {
...
+            $prefix = "INSERT INTO " . table($table) . " (" . implode(", ", $columns) . ") VALUES\n";
+            $values = array();
+            foreach ($columns as $key) {
+                $values[$key] = "$key = VALUES($key)";
+            }
+            $suffix = "\nON DUPLICATE KEY UPDATE " . implode(", ", $values);
+            $values = array();
+            $length = 0;
+            foreach ($rows as $set) {
+                $q = "MERGE INTO ".$table." d
```
Comment (vrana) [inline adminer/drivers/oracle.inc.php]: Quote `$table` using `table()`.

File: adminer/drivers/oracle.inc.php [hunk 4]
```diff
@@ -136,6 +136,41 @@ function select_db($database) {
...
+                    USING (SELECT ".$primary." from ".$table.") s
```
Comment (vrana) [inline adminer/drivers/oracle.inc.php]: `$primary` is an array. What's this supposed to do?
Quote `$primary`.

Comment (vrana) [inline adminer/drivers/oracle.inc.php]: Use uppercase for keywords.

Comment (vrana) [inline adminer/drivers/oracle.inc.php]: Avoid indenting SQL queries just because Adminer's code is indented.

File: adminer/drivers/oracle.inc.php [hunk 5]
```diff
@@ -158,14 +193,14 @@ function connect() {
 		global $adminer;
 		$connection = new Min_DB;
 		$credentials = $adminer->credentials();
-		if ($connection->connect($credentials[0], $credentials[1], $credentials[2])) {
+		if ($connection->connect(DB, $credentials[1], $credentials[2])) {
```
Comment (vrana) [inline adminer/drivers/oracle.inc.php]: Why this change?

File: adminer/drivers/oracle.inc.php [hunk 6]
```diff
@@ -158,14 +193,14 @@ function connect() {
...
 			return $connection;
 		}
 		return $connection->error;
 	}
 
 	function get_databases() {
-		return get_vals("SELECT tablespace_name FROM user_tablespaces");
+		return get_vals("select sys_context('userenv','instance_name') from dual");
```
Comment (vrana) [inline adminer/drivers/oracle.inc.php]: Add space after comma.

File: adminer/drivers/oracle.inc.php [hunk 7]
```diff
@@ -194,7 +229,7 @@ function logged_user() {
 	}
 
 	function tables_list() {
-		return get_key_vals("SELECT table_name, 'table' FROM all_tables WHERE tablespace_name = " . q(DB) . "
+		return get_key_vals("SELECT table_name, 'table' FROM all_tables WHERE owner = " .q(get_schema()). "
```
Comment (vrana) [inline adminer/drivers/oracle.inc.php]: Surround concatenation (`.`) with spaces.

File: adminer/include/functions.inc.php [hunk 8]
```diff
@@ -405,14 +405,19 @@ function where($where, $fields = array()) {
...
 		$column = escape_key($key);
-		$return[] = $column
-			. (($jush == "sql" && preg_match('~^[0-9]*\\.[0-9]*$~', $val)) || $jush == "mssql"
-				? " LIKE " . q(addcslashes($val, "%_\\"))
-				: " = " . unconvert_field($fields[$key], q($val))
-			) // LIKE because of floats but slow with ints, in MS SQL because of text
-		; //! enum and set
-		if ($jush == "sql" && preg_match('~char|text~', $fields[$key]["type"]) && preg_match("~[^ -@]~", $val)) { // not just [a-z] to catch non-ASCII characters
-			$return[] = "$column = " . q($val) . " COLLATE " . charset($connection) . "_bin";
+		if ($jush == "oracle" && $fields[$key]["type"] == "clob") {
```
Comment (vrana) [inline adminer/include/functions.inc.php]: Minimize changes in common files, move them to the driver if possible. (I know that this code already violates this rule but don't make it worse.)
